# leetcode 33: Search in Rotated Sorted Array


In [27]:
def search(nums, target):
    """
    :type nums: List[int]
    :type target: int
    :rtype: int
    """
    print("nums: ", nums, "target: ", target)

    left = 0
    right = len(nums) - 1
    i = 0

    while left <= right:

        # avoid overflow
        mid = left + (right - left) // 2

        if target == nums[mid]:
            return mid
        # one part must be sorted
        if nums[left] <= nums[mid]:
            # if right half is sorted
            if nums[left] <= target and target <= nums[mid]:
                print("key is in left half")

                right = mid - 1

            else:
                left = mid + 1

        else:
            # if right half is sorted

            print("right - half is sorted")
            if nums[mid] <= target and target <= nums[right]:
                # if key is in left half
                left = mid + 1
            else:
                right = mid - 1

    return -1


nums = [4, 5, 6, 7, 0, 1, 2]
search(nums, target=2)

nums:  [4, 5, 6, 7, 0, 1, 2] target:  2


6

In [28]:
def search(nums, target):
    left, right = 0, len(nums) - 1

    while left <= right:
        mid = left + (right - left) // 2
        if nums[mid] == target:
            return mid

        # one of side must be sorted

        # bisecting fo rbinary tree search
        if nums[left] <= nums[mid]:
            #

            # if target is on left
            if nums[left] <= target <= nums[mid]:
                right = mid - 1
            else:
                left = mid + 1  # switch to other side
        else:

            if nums[mid] <= target <= nums[right]:
                left = mid + 1
            else:
                right = mid - 1

    return -1


nums = [4, 5, 6, 7, 0, 1, 2]
search(nums, target=2)

6

# Refined


In [29]:
def search(nums, target):
    print(f"nums: {nums}, target:{target}")

    left = 0
    right = len(nums) - 1

    while left <= right:
        mid = left + (right - left) // 2
        print(f"left: {left}, right:{right}, nums:mid:{nums[mid]}")
        if nums[mid] == target:
            return mid

        # left side
        if target <= nums[mid]:
            print(f"going to left bcz target is less than/equal to middle")

            if nums[left] <= target <= nums[mid]:
                print("confirmed: target is in left part and left part is ordered")
                right = mid - 1
            else:
                print("shifting: target is in right part")
                left = mid + 1

        else:
            print(f"going to right bcz target is greater than middle")

            if nums[mid] <= target <= nums[right]:
                print("confirmed: target is in right part and right part is ordered")
                left = mid + 1
            else:
                print("shifting: target is in left part")
                right = mid - 1

    return -1


nums = [4, 5, 6, 7, 0, 1, 2]
search(nums, target=2)

nums: [4, 5, 6, 7, 0, 1, 2], target:2
left: 0, right:6, nums:mid:7
going to left bcz target is less than/equal to middle
shifting: target is in right part
left: 4, right:6, nums:mid:1
going to right bcz target is greater than middle
confirmed: target is in right part and right part is ordered
left: 6, right:6, nums:mid:2


6

In [30]:
s = "dfa12321afd"
s.split()


for i in s:
    # print(i)
    if i.isdigit():
        print(i)

1
2
3
2
1


## 1796. Second Largest Digit in a String


In [31]:
s = "abc1123"
numeric = [int(i) for i in s if i.isdigit()]


max = -1
secondMax = -1
for n in numeric:
    if n > max:
        secondMax = max
        max = n
    elif n > secondMax and n < max:
        secondMax = n
print(secondMax)

2


## 283. Move Zeroes


In [32]:
# swap(nums, 1, 3)

In [33]:
nums = [0, 0, 1, 0, 0, 3, 12]


def move_zeros_to_end(arr):
    non_zero_pos = 0  # Position to place the next non-zero element

    # Iterate through the array
    for current in range(len(arr)):
        if arr[current] != 0:
            # Swap the elements at non_zero_pos and current
            arr[non_zero_pos], arr[current] = arr[current], arr[non_zero_pos]
            non_zero_pos += 1  # Move the non-zero position to the right

    print(arr)


move_zeros_to_end(nums)

[1, 3, 12, 0, 0, 0, 0]


In [34]:
# nums = [0, 0, 1, 0, 0, 3, 12]


# non_zero_pos = 0


# for current in range(len(nums)):

#     print(
#         f"idx: {current}, non_zero_pos: {nums[non_zero_pos]}, current: {nums[current]}"
#     )

#     if nums[current] != 0:
#         # print(current)
#         print(f"non_zero_pos: {nums[non_zero_pos]}, current: {nums[current]}")
#         print("\t entered")

#         print(f"\t non_zero_pos: {nums[non_zero_pos]}, current: {nums[current]}")
#         nums[non_zero_pos], nums[current] = nums[current], nums[non_zero_pos]
#         non_zero_pos += 1

# print(nums)

In [35]:
nums = [0, 0, 1, 0, 0, 3, 12]

left = 0  # permanent
current = 0


while current < len(nums):
    print(nums[current])

    if nums[current] != 0:
        nums[left], nums[current] = nums[current], nums[left]
        left += 1

    current += 1


print(nums)

0
0
1
0
0
3
12
[1, 3, 12, 0, 0, 0, 0]


In [36]:
def moveZeroes(nums):
    """
    :type nums: List[int]
    :rtype: None Do not return anything, modify nums in-place instead.
    """
    zero_count = nums.count(0)
    l = 0
    for n in nums:
        if n != 0:
            nums[l] = n
            l += 1
    for zero in range(1, zero_count + 1):
        nums[-zero] = 0

    return nums


moveZeroes(nums)

[1, 3, 12, 0, 0, 0, 0]

# 7: Given a signed 32-bit integer x, return x with its digits reversed. If reversing x causes the value to go outside the signed 32-bit integer range [-231, 231 - 1], then return 0.


In [37]:
def reverseInt(x):
    sign = 1 if x > 0 else -1
    reversed_no = 0
    x = abs(x)

    while x > 0:
        digit = x % 10
        x = x // 10
        reversed_no = reversed_no * 10 + digit  # 32 = 3 *10 +2

    return sign * reversed_no


x = -123

reverseInt(x)

-321

In [38]:
def reverseInt(x):

    MAX_INT = 2**31 - 1  # 2,147,483,647
    MIN_INT = -(2**31)  # -2,147,483,648

    sign = 1 if x > 0 else -1
    reversed_no = 0
    x = abs(x)

    while x > 0:
        if (
            reversed_no > MAX_INT / 10 or reversed_no < MIN_INT / 10
        ):  # MIN_INT / 10: not including units place
            return 0
        digit = x % 10
        x = x // 10
        reversed_no = reversed_no * 10 + digit

    return sign * reversed_no


x = -123123123312313123

reverseInt(x)

0

# revserse an array


In [39]:
def reverseArray(array):
    left = 0
    right = len(nums) - 1
    while left < right:
        nums[left], nums[right] = nums[right], nums[left]
        left += 1
        right -= 1

    return nums


nums = [1, 2, 3]
reverseArray(nums)

[3, 2, 1]

In [40]:
s = "sre"

s[::-1]

'ers'

In [41]:
my_list = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]


my_list[0:0:-1]

[]

## 268. Missing Number

Given an array nums containing n distinct numbers in the range [0, n], return the only number in the range that is missing from the array.


In [42]:
def missingNumber(nums):
    n = len(nums)
    expected_sum = n * (n + 1) // 2
    return expected_sum - sum(nums)


nums = [9, 6, 4, 2, 3, 5, 7, 0, 1]

missingNumber(nums)

8

In [43]:
def gcdOfStrings(str1: str, str2: str) -> str:
    # Check if concatenated strings are equal or not, if not return ""
    if str1 + str2 != str2 + str1:
        return ""
    # If strings are equal than return the substring from 0 to gcd of size(str1), size(str2)
    from math import gcd

    print(f"s1: {len(str1)}, s2: {len(str2)}")
    return str1[: gcd(len(str1), len(str2))]

In [44]:
import math
from math import gcd

gcd(10, 12)

2

In [45]:
gcdOfStrings("ABABAB", "AB")

s1: 6, s2: 2


'AB'

# sumsTwo


In [46]:
nums = [16, 4, 23, 8, 15, 42, 1, 2]
target = 19

In [47]:
nums.sort()
print(nums)

[1, 2, 4, 8, 15, 16, 23, 42]


In [48]:
p = 0
q = len(nums) - 1

In [49]:
nums[q]

42

In [50]:
# if( nums[p] + nums[q] > target) and nums[q] > target :
#     q-=1

In [51]:
while nums[p] + nums[q] != target:
    #     if( nums[p] + nums[q] > target) and nums[q] > target :
    #         q-=1
    #     if( nums[p] + nums[q] > target) and nums[q] < target :
    #         p+=1
    if nums[p] + nums[q] > target:
        q -= 1

    if nums[q] > target:
        q -= 1
    if nums[q] < target:
        p += 1

    print(f"{nums[p]} + {nums[q]}  = {nums[p] + nums[q]}")

2 + 16  = 18
4 + 16  = 20
8 + 15  = 23
15 + 8  = 23
16 + 4  = 20
23 + 2  = 25
42 + 1  = 43
42 + 23  = 65


IndexError: list index out of range

In [ ]:
if (nums[p] + nums[q] > target) and p < 10:
    p += 1

In [ ]:
dict = {}

In [ ]:
nums = [16, 4, 23, 8, 15, 42, 1, 2]
target = 58

In [ ]:
nums.sort()
print(nums)

In [ ]:
def fn(nums):
    for i, val in enumerate(nums):
        print([(i, j) for i, j in dict.items()])
        if (target - val) in dict.keys():
            return (i, dict[target - val])
        else:
            dict[val] = i


fn(nums)

In [ ]:
%%time

In [ ]:
nums = [16, 4, 23, 8, 15, 42, 1, 2]
target = 58
nums.sort()
print(nums)

In [ ]:
%%time


def fn(nums):
    dic=dict()

    for i in range(len(nums)):
        if( (target - nums[i]) in dic.keys() ):
            return (dic[target - nums[i]], i)
        else:
            dic[nums[i]] = i
fn(nums)        

In [ ]:
%%time


def fn(nums):
    dic=dict()

    for i in range(len(nums)):

        rem = target - nums[i]
        if( (rem) in dic.keys() ):
            return (dic[rem], i)
        else:
            dic[nums[i]] = i
fn(nums)        

In [ ]:
%%time
prices = [7,1,5,3,6,4]
prices =[1,2]
def maxProfit(prices):
    buy = prices.index(min(prices))
#     print(f"BUY| day:{buy} , value:{prices[buy]}")
    maxProfit = max(prices[buy:])
    sell =  prices.index(maxProfit)+1 if(maxProfit>prices[buy])  else 0   # +1 for 0 indexing
#     print(f"SELL| day:{sell} , value:{prices[sell]}")
    
#     
    return sell

maxProfit(prices) 

In [ ]:
%%time
prices = [7,1,5,3,6,4]
prices =[1,2]
def maxProfit(prices):
    buy = prices.index(min(prices))
    print(f"BUY| day:{buy} , value:{prices[buy]}")
    sell = prices.index(max(prices[buy:]))
#     sell =  prices.index(maxProfit)+1 if(maxProfit>prices[buy])  else 0   # +1 for 0 indexing
    print(buy,sell)
    profit = prices[sell] - prices[buy]
    return profit if(profit) > 0 else 0
#     print(f"SELL| day:{sell} , value:{prices[sell]}")
    
#     
#     return sell

maxProfit(prices) 

In [ ]:
# from memory_profiler import profile
# @profile(precision=4)


prices = [7, 1, 5, 3, 6, 4]
prices = [1, 2]
margin = 0
left = 0
right = 1

while right < len(prices):
    if prices[right] < prices[left]:  # buy at min
        left = right
    if prices[right] >= prices[left]:
        margin = (
            prices[right] - prices[left]
            if prices[right] - prices[left] > margin
            else margin
        )

        right += 1
margin

In [ ]:
# from memory_profiler import profile
# @profile(precision=4)


prices = [7, 1, 5, 3, 6, 4]
prices = [1, 2]
margin = 0
left = 0
right = 1

while right < len(prices):
    if prices[right] > prices[left]:
        margin = max(margin, prices[right] - prices[left])
    else:
        left += 1
    right += 1
margin

# X


In [ ]:
txt = "A shallow copy creates a new object, but it does not create copies of the objects that the original object referencesddd."
txt = txt.lower().replace(',', "").replace('.', "").replace(' ', "")

In [ ]:
alphabets = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z']

for i, c in enumerate(alpahabets):
    print(f"char:{c}, index:{i}, count:{txt.count(c)}") if txt.count(c) == i+1 else ""
    # print(f"c:{c}, i:{i}, count:{txt.count(c)}")
    

In [ ]:
txt = "A shallow copy creates a new object, but it does not create copies of the objects that the original object referencesddd."
txt = txt.lower().replace(',', "").replace('.', "").replace(' ', "")
alphabets = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z']


dict={}
for i, c in enumerate(alphabets):
    dict[c] = txt.count(c) if txt.count(c) == i+1 else None
    # print(f"c:{c}, i:{i}, count:{txt.count(c)}")
print(dict)

In [ ]:
txt = "A shallow copy creates a new object, but it does not create copies of the objects that the original object referencesddd."
alphabets = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z']
txt = txt.lower().replace(',', "").replace('.', "").replace(' ', "")
{ c: txt.count(c) for i,c in enumerate(alphabets) if txt.count(c) == i+1}


In [ ]:
ord('a')

In [ ]:
chr(97)

#


In [ ]:
# final
alphabets = [chr(i) for i in range(ord('a'), ord('z')+1)]
txt = "A shallow copy creates a new object, but it does not create copies of the objects that the original object referencesddd."
txt = txt.lower().replace(',', "").replace('.', "").replace(' ', "")
{ c: txt.count(c) for i,c in enumerate(alphabets) if txt.count(c) == i+1}


# 451. Sort Characters By Frequency


In [ ]:
txt= "treefffftzzzzz"

unique = set(txt)
unique

# x = { s.count(char) for char in unique} # set

dic = { char:txt.count(char) for char in unique} # dict
dic

In [ ]:
# sorted(dic,  key = lambda item: dic[item])  # sorted but only keys



In [ ]:
sorted_dic = sorted(dic.items(), key = lambda item: item[1], reverse = True)
sorted_dic = dict(sorted_dic)
sorted_dic

In [ ]:
result = ""
for index, value in sorted_dic.items():
    # print(f"index: {index} value: {value}")
    result += index*value
print(result) 

In [ ]:
''.join(['1','2','3'])

In [ ]:
''.join(sorted_dic.keys())

In [ ]:
# final
txt= "treefffftzzzzz"
# result = ""

unique = set(txt)
dic = { char:txt.count(char) for char in unique} # dict
sorted_dic = sorted(dic.items(), key = lambda item: item[1], reverse = True)
sorted_dic = dict(sorted_dic)
# for index, value in sorted_dic.items():
#     result += index*value
''.join(key*value for key, value in sorted_dic.items())
print(result) 

In [ ]:
dic

In [ ]:
# final
txt= "tree"

unique = set(txt)
dic = { char:txt.count(char) for char in unique} # dict
sorted_dic = sorted(dic.items(), key = lambda item: item[1], reverse = True)
print(sorted_dic)
# sorted_dic = dict(sorted_dic)
# ''.join(key*value for key, value in sorted_dic.items())
''.join(char*count for char, count in sorted_dic)


In [ ]:
class Solution(object):
    def frequencySort(self, s):
        """
        :type s: str
        :rtype: str
        """
        unique = set(s)
        dic = { char:s.count(char) for char in unique} # dict
        sorted_dic = sorted(dic.items(), key = lambda item: item[1], reverse = True)
        s = ''.join(char*count for char, count in sorted_dic)
        return ''.join(char*count for char, count in sorted_dic)



# 647. Palindromic Substrings


In [ ]:

str= "abcde"

# for i in range(len("abcde")//2):
i=0
while(i< len(str)//2):
      print(i)
      
      i+=1

In [ ]:
round(len("abc"))//2, len("abc")//2

In [ ]:
def palindrome(s, i=0, j=None):
    
    if j is None:
        j=len(s)-1
    print(f"s:{s}, i:{i}, j:{j}, s[i]:{s[i]}, s[j]:{s[j]}")
    if(i>=j):
        return True
    if(s[i] != s[j]): 
        return False
    # i+=1;j-=1
    # print(s, i, j)
    return palindrome(s, i+1, j-1)   # necessary to return True   otherwise result will be lost
 
s = "abcba"
palindrome(s)

# Hckerrank: breaking the recored

In [ ]:
min_scores = max_scores = scores[0]
min_scores ,max_scores


count_min_record_broke = 0
count_max_record_broke = 0

In [ ]:
scores = [10, 5, 20, 20, 4, 5, 2, 25, 1]

min_scores = max_scores = scores[0]



count_min_record_broke = 0
count_max_record_broke = 0


for index in range(len(scores)):

    if scores[index]<min_scores:
        min_scores = scores[index]
        count_min_record_broke+=1
    if scores[index]>max_scores:
        max_scores = scores[index]
        count_max_record_broke+=1

    print(f"index: {index}, min_scores: {min_scores} , max_scores: {max_scores}")

count_max_record_broke, count_min_record_broke

In [ ]:
scores = [10, 5, 20, 20, 4, 5, 2, 25, 1]

min_scores = max_scores = scores[0]

count_min_record_broke = 0
count_max_record_broke = 0

for index, score in enumerate(scores):
    if score < min_scores:
        min_scores = score
        count_min_record_broke += 1
    elif score > max_scores:  # Use 'elif' since a score can't be both min and max in the same iteration
        max_scores = score
        count_max_record_broke += 1

    print(f"index: {index}, min_scores: {min_scores}, max_scores: {max_scores}")

count_max_record_broke, count_min_record_broke


# 739. Daily Temperatures


Given an array of integers temperatures represents the daily temperatures, return an array answer such that answer[i] is the number of days you have to wait after the ith day to get a warmer temperature. If there is no future day for which this is possible, keep answer[i] == 0 instead.

 

Example 1:

Input: temperatures = [73,74,75,71,69,72,76,73]

Output: [1,1,4,2,1,1,0,0]

Example 2:

Input: temperatures = [30,40,50,60]

Output: [1,1,1,0]

Example 3:

Input: temperatures = [30,60,90]

Output: [1,1,0]


In [ ]:
# daily_tempratures = [73,74,75,71,69,72,76,73]

days_to_wait= ["0"] * len(daily_tempratures)

for dailyIndex, dailyT in enumerate(daily_tempratures):

    for index in range(dailyIndex , len(daily_tempratures)):
        print(f"index: {index}, dailyIndex:{dailyIndex}")



In [ ]:
def next_hotter(lst, ind):
    # print(lst[0])
    current = lst[ind]
    for i in range(ind, len(lst)):
        if(current< lst[i]):
            # print(f"current:{current}, i:{i}, lst[i]:{lst[i]}")
            hotter = i-ind
            print(f"current:{current}, i:{i}, lst[i]:{lst[i]}, hotter:{hotter}")
            return hotter



daily_tempratures = [73,70,49,74,75]
print(daily_tempratures)
next_hotter(daily_tempratures,0)

In [ ]:
def next_hotter_whole(lst):
    changes = [0] * len(lst)
    # print(changes)
    # print(lst[0])
    for index, value in enumerate(lst):
        print(f"\nindex:{index}, value:{value}")

        for i in range(index, len(lst)):
            if(value < lst[i]):
                # print(f"current:{current}, i:{i}, lst[i]:{lst[i]}")
                # hotter = i-index-1
                print(f"\t\t\tcurrent:{value}, i:{i}, lst[i]:{lst[i]}, hotter:{i-index}")
                changes[index] = i-index
                break;

    return changes

                



daily_tempratures = [73,74,75,71,69,72,76,73]
print(daily_tempratures)
next_hotter_whole(daily_tempratures)

In [ ]:
daily_tempratures



In [ ]:
daily_tempratures = [73, 74, 75, 71, 69, 72, 76, 73]


st=[]

res = [0] * len(daily_tempratures)
 
for i in range(len(daily_tempratures)):

    while st and daily_tempratures[i]  > daily_tempratures[st[-1]]:
        print(f"\t\tdaily_tempratures[i]: {daily_tempratures[i]}, daily_tempratures[st[-1]]: {daily_tempratures[st[-1]]}, \t\tst: {st}")
        idx = st.pop()
        res[idx] = i - idx
        print(f"\t\tidx:pop():{idx}, {daily_tempratures[idx]}\t||, res:{res}")
    st.append(i)
    
    print("status :st:",st)
    print("--")


In [ ]:
st

In [ ]:
st.pop()

In [ ]:
st

In [ ]:
st[1], st[-1]

# sorting using stack


In [ ]:


input_stack = [1,2,3,4,5]
input_stack = sorted(input_stack, reverse=True)
sorted_stack =[] 


while(input_stack):
    current = input_stack.pop()

    while sorted_stack and sorted_stack[-1] > current:
        input_stack.append(sorted_stack.pop())
    sorted_stack.append(current)
    # print(current)


    print(f"Status:\n\t\tinput_stack:{input_stack}, sorted_stack:{sorted_stack}, current:{current}")



# while()

In [ ]:


input_stack = [1,2,3,4,5]
input_stack = sorted(input_stack, reverse=True)
sorted_stack =[] 


while(input_stack):
    current = input_stack.pop()

    while sorted_stack and sorted_stack[-1] > current:
        input_stack.append(sorted_stack.pop())
    sorted_stack.append(current)
    # print(current)


    print(f"Status:\n\t\tinput_stack:{input_stack}, sorted_stack:{sorted_stack}, current:{current}")



# while()

# 739. Daily Temperatures - with stack


In [ ]:

# final
daily_tempratures = [73, 74, 75, 71, 69, 72, 76, 73]

print(f"daily_tempratures: {daily_tempratures}")

st=[]

res = [0] * len(daily_tempratures)

for i in range(len(daily_tempratures)):
    current = daily_tempratures[i]
    # print(f"\tcurrent: {current}, i:{i}")
    while(st and daily_tempratures[i]>daily_tempratures[st[-1]]):# current bigger than top of stack
        # print(f"Status:\t\tst:{st}, curret:{ daily_tempratures[i]}, top of stack:{daily_tempratures[st[-1]]}")
        idx = st.pop()  # diifernce between 
        difference = i - idx
        res[idx] = difference

    st.append(i)

print(res)




daily_tempratures: [73, 74, 75, 71, 69, 72, 76, 73]
[1, 1, 4, 2, 1, 1, 0, 0]


In [ ]:
# optimized
answer = [0] * len(daily_tempratures)
stack = []

for index, current_temp in enumerate(daily_tempratures):
    while stack and daily_tempratures[stack[-1]] < current_temp:
        previous_index = stack.pop()
        answer[previous_index] = index - previous_index
    stack.append(index)
answer

[1, 1, 4, 2, 1, 1, 0, 0]

In [ ]:

# final
daily_tempratures = [73, 74, 75, 71, 69, 72, 76, 73]

print(f"daily_tempratures: {daily_tempratures}")
st=[]
res = [0] * len(daily_tempratures)
for index, current in enumerate(daily_tempratures):
    while(st and current>daily_tempratures[st[-1]]):# current bigger than top of stack
        prev_index = st.pop()  # diifernce between 
        res[prev_index] = index - prev_index
    st.append(index)

print(res)




daily_tempratures: [73, 74, 75, 71, 69, 72, 76, 73]
[1, 1, 4, 2, 1, 1, 0, 0]


# Camel Case: 04

In [ ]:
def splitting(input_txt_name, input_txt_type):
        print("We'll be spliting...", input_txt_name)

        names = []
        prev_index = 0
        length = len(input_txt_name)
        for current in range(length):
              if input_txt_name[current].isupper():
                    names.append(input_txt_name[prev_index:current].lower())
                    # print(input_txt_name[prev_index:current].lower())
                    prev_index = current
                    
        
        # print("input_txt_type", input_txt_type)
        
        lastportion = input_txt_name[prev_index:-2] if input_txt_type == 'M' else input_txt_name[prev_index:]
        names.append(lastportion.lower())
        return " ".join(names).strip()
        



def combining(input_txt_name, input_txt_type):
        print("We'll be combining...", input_txt_name)
        input_txt_name = input_txt_name.split(" ")
        print(input_txt_name)
        


        # for current in input_txt_name:
        #       current = curr

        for current in range(len(input_txt_name)):
            input_txt_name[current] = input_txt_name[current].title()
        
        print(input_txt_name)
        name = "".join(input_txt_name)
        
        
        if input_txt_type == 'C':
            return name
        return name.lower() + "()" if input_txt_type == 'M' else name.lower() 
        
        


def parsingInputString(input_txt):
    print(f"input_text: {input_txt}")


    input_txt_operation = input_txt[0:1]
    print(f"input_txt_operation: {input_txt_operation}")

    input_txt_type = input_txt[2:3]
    print(f"input_txt_type: {input_txt_type}")

    input_txt_name = input_txt[4:]
    print(f"input_txt_name: {input_txt_name}")



    if input_txt_operation == 'C':
        return combining(input_txt_name, input_txt_type)
    if input_txt_operation == 'S':
        return splitting(input_txt_name, input_txt_type)
    


# S;M;plasticCup()
# C;V;mobile phone
# C;C;coffee machine
# S;C;LargeSoftwareBook
# C;M;white sheet of paper




# input_txt =  "S;V;pictureFrameNice"
# input_txt = "S;M;plasticCup()"
# input_txt = "S;V;pictureFrame"
# input_txt = "S;C;LargeSoftwareBook"
# input_txt = "S;V;pictureFrame"


# input_txt = "C;V;mobile phone" # mobilePhone
input_txt = "C;C;coffee machine" # CoffeeMachine
# input_txt = "C;M;white sheet of paper" # whiteSheetOfPaper()





parsingInputString(input_txt)

input_text: C;C;coffee machine
input_txt_operation: C
input_txt_type: C
input_txt_name: coffee machine
We'll be combining... coffee machine
['coffee', 'machine']
['Coffee', 'Machine']


'CoffeeMachine'

In [ ]:
def splitting(input_txt_name, input_txt_type):
        print("We'll be spliting...", input_txt_name)

        names = []
        prev_index = 0
        length = len(input_txt_name)
        for current in range(length):
              if input_txt_name[current].isupper():
                    names.append(input_txt_name[prev_index:current].lower())
                    prev_index = current
                    
        lastportion = input_txt_name[prev_index:-2] if input_txt_type == 'M' else input_txt_name[prev_index:]
        names.append(lastportion.lower())
        return " ".join(names).strip()
        

def combining(input_txt_name, input_txt_type):
        input_txt_name = input_txt_name.split(" ")
        for current in range(len(input_txt_name)):
            input_txt_name[current] = input_txt_name[current].title()
        
        print(input_txt_name)
        
        
        
        if input_txt_type == 'C':
            return "".join(input_txt_name)
        

        input_txt_name[0] = input_txt_name[0].lower()
        name = "".join(input_txt_name)
        # print(name)

        # new_name = name[0].lower + name[1:]
        # print(new_name)
        
        return name + "()" if input_txt_type == 'M' else name() 
        

def parsingInputString(input_txt):
    input_txt_operation = input_txt[0:1]
    input_txt_type = input_txt[2:3]
    input_txt_name = input_txt[4:]
    if input_txt_operation == 'C':
        return combining(input_txt_name, input_txt_type)
    if input_txt_operation == 'S':
        return splitting(input_txt_name, input_txt_type)


input_txt = "C;M;mouse pad"
 # CoffeeMachine





parsingInputString(input_txt)

['Mouse', 'Pad']
mousePad


'mousePad()'

In [ ]:
# for hackerrank 

import sys



def splitting(input_txt_name, input_txt_type):
        
        names = []
        prev_index = 0
        length = len(input_txt_name)
        for current in range(length):
              if input_txt_name[current].isupper():
                    names.append(input_txt_name[prev_index:current].lower())
                    prev_index = current
                    
        lastportion = input_txt_name[prev_index:-2] if input_txt_type == 'M' else input_txt_name[prev_index:]
        names.append(lastportion.lower())
        name =  " ".join(names).strip()
        print(name)
        return name
        

def combining(input_txt_name, input_txt_type):
        input_txt_name = input_txt_name.split(" ")
        for current in range(len(input_txt_name)):
            input_txt_name[current] = input_txt_name[current].title()
        
        
        
        if input_txt_type == 'C':
            name = "".join(input_txt_name)
            print(name)
            return name
        
        input_txt_name[0] = input_txt_name[0].lower()
        name = "".join(input_txt_name)
        name = name + "()" if input_txt_type == 'M' else name()
        print(name)
        return name
        

def parsingInputString(input_txt):
    input_txt_operation = input_txt[0:1]
    input_txt_type = input_txt[2:3]
    input_txt_name = input_txt[4:]
    if input_txt_operation == 'C':
        return combining(input_txt_name, input_txt_type)
    if input_txt_operation == 'S':
        return splitting(input_txt_name, input_txt_type)





for line in sys.stdin:
    line = line.strip()
    parsingInputString(line)

KeyboardInterrupt: 

In [ ]:
# submiited 

import sys

def splitting(input_txt_name, input_txt_type):
    names = []
    start = 0
    for i, char in enumerate(input_txt_name):
        if char.isupper() and i > 0:
            names.append(input_txt_name[start:i].lower())
            start = i
    
    last_part = input_txt_name[start:-2] if input_txt_type == 'M' else input_txt_name[start:]
    names.append(last_part.lower())
    
    name = " ".join(names).strip()
    print(name)
    return name

def combining(input_txt_name, input_txt_type):
    words = input_txt_name.split()
    if input_txt_type == 'C':
        name = "".join(word.capitalize() for word in words)
    else:
        print(words)
        name = words[0].lower() + "".join(word.capitalize() for word in words[1:])
        name += "()" if input_txt_type == 'M' else ""
    
    print(name)
    return name

def parsingInputString(input_txt):
    input_txt_operation = input_txt[0]
    input_txt_type = input_txt[2]
    input_txt_name = input_txt[4:]
    
    if input_txt_operation == 'C':
        return combining(input_txt_name, input_txt_type)
    if input_txt_operation == 'S':
        return splitting(input_txt_name, input_txt_type)



input_txt = "C;M;mouse pad"
 # CoffeeMachine





parsingInputString(input_txt)

['mouse', 'pad']
mousePad()


'mousePad()'

In [ ]:
"mousepad".capitalize()

'Mousepad'

In [ ]:
"mousepad".title()

'Mousepad'

# hackerrank:  divisble sums pairs

In [ ]:
arr = [1, 3, 2, 6, 1, 2]
n = 6; k = 3
arr

[1, 3, 2, 6, 1, 2]

In [ ]:

# arr = sorted(arr)

count =0
for index, value in enumerate(arr):
    for i in range(index+1, len(arr)):
        # print("i:{0}, value:{1}, index:{2}".format( i, value, index))

        if ( True and (value + arr[i])%k==0 ):
            # print(value, arr[i])
            count+=1


print(count)

5


# Sparse arrays

There is a collection of input strings and a collection of query strings. For each query string, determine how many times it occurs in the list of input strings. Return an array of the results.


In [ ]:
# strings = ['aba' ,'baba' ,'aba' ,'xzxb']
# query = ['aba', 'xzxb', 'ab']

strings = ['abcde','sdaklfj','asdjf','na','basdn','sdaklfj','asdjf','na','asdjf','na','basdn','sdaklfj','asdjf']
query = ['abcde','sdaklfj','asdjf','na','basdn']
count = [0] * len(query)

strings, query, count

(['abcde',
  'sdaklfj',
  'asdjf',
  'na',
  'basdn',
  'sdaklfj',
  'asdjf',
  'na',
  'asdjf',
  'na',
  'basdn',
  'sdaklfj',
  'asdjf'],
 ['abcde', 'sdaklfj', 'asdjf', 'na', 'basdn'],
 [0, 0, 0, 0, 0])

In [ ]:
for qindex, query in enumerate(query):

    for string in strings:
        if query == string:
            count[qindex]+=1


count

[1, 3, 4, 3, 2]

# 347. Top K Frequent Elements
Given an integer array nums and an integer k, return the k most frequent elements. You may return the answer in any order.

Input: nums = [1,1,1,2,2,3], k = 2
Output: [1,2]

In [ ]:
# nums = [1,1,1,2,2,3,3,3,3,3]; k = 2
# Output: [1,2]
nums = [-1,2,2] 
k=1

result=[]

for index, value in enumerate(set(nums)):
    result.append(nums.count(value))
    print(result)
# result = sorted(result)
print(result)
result[:k]

[2]
[2, 1]
[2, 1]


[2]

In [ ]:
# nums = [1,1,1,2,2,3,3,3,3,3]; k = 2
# Output: [1,2]
nums = [1,1,1,2,2,3]
k=2

result={}

for value in set(nums):
    # result.append(nums.count(value))
    result[value] = nums.count(value)
    # print(result)
# result = sorted(result)
print(result)


sorted_result = sorted(result.items(), key=lambda item: item[1], reverse=True)
sorted_result




sorted_keys = [ keys for keys, v in sorted_result]
sorted_keys[:k]


{1: 3, 2: 2, 3: 1}


[1, 2]

In [ ]:
# nums = [1,1,1,2,2,3,3,3,3,3]; k = 2
# Output: [1,2]

#beats 86%
nums = [1,1,1,2,2,3]
k=2

result={}


for number in nums:
    if number in result:
        result[number] +=1
    else:
        result[number] = 1


print(result)

sorted_result = sorted(result.items(), key=lambda item: item[1], reverse=True)
sorted_result




sorted_keys = [ keys for keys, v in sorted_result]
sorted_keys[:k]


{1: 3, 2: 2, 3: 1}


[1, 2]

In [ ]:
"aabc".count("a")

2

In [ ]:
# optimal 
import heapq
freq_map = {}
for num in nums:
    if num in freq_map:
        freq_map[num] += 1
    else:
        freq_map[num] = 1

# Step 2: Use a heap to find the k most frequent elements
# The heap will store elements as (-frequency, element) to create a max heap
heap = []
for num, freq in freq_map.items():
    heapq.heappush(heap, (-freq, num))

# Step 3: Extract the top k elements from the heap
top_k = []
for _ in range(k):
    top_k.append(heapq.heappop(heap)[1])

top_k

[1, 2]

# String Encode and Decode

Input: ["neet","code","love","you"]

Output:["neet","code","love","you"]


In [ ]:
input = [""]

encoded = "#".join(input)
print("encoded", encoded)
# decoded = encoded.split("#") if encoded!="" else []

decoded = encoded.split("#")
print(decoded)



encoded 
['']


In [ ]:
"#".join([""]), "#".join([])

('', '')

### (len)#(string)

In [ ]:
input = ["neet","code","love","you"]
# input=[]
string=""
for i in input:
    string+=str(len(i))+'#' + i
string


'4#neet4#code4#love3#you'

In [ ]:

output =[]
i=0
while(i<len(string)):
    length = int(string[i])
    print("length:", length)
    print(string[i+1+1:i+1+1+length])
    output.append(string[i+1+1:i+1+1+length])
    i = i+ 1+length+1

output


length: 4
neet
length: 4
code
length: 4
love
length: 3
you


['neet', 'code', 'love', 'you']

In [ ]:
# accepted
# input = ["neet","code","love","you"]
input=[""]
string=""
for i in input:
    string+=str(len(i))+'#' + i
string

print(string)
res =[]
i=0
while(i<len(string)):
    j=i
    while(string[j]!='#'):
        j+=1
    length=int(string[i:j])

    print(string[j+1:j+1+length])
    res.append(string[j+1:j+1+length])
    i = j+1+length
res



0#



['']

In [ ]:

Input:

[""]

Your Output:

[]

Expected output:

[""]

## median

In [ ]:
arr  = [5,1,2,3,4,6]

arr = sorted(arr)
arr
length =len(arr)

if length%2 ==0:
    print(length//2)
    print((arr[length//2 -1] +  arr[length//2]) / 2)

else:
    print((length//2))





3
3.5


In [ ]:
# fina
arr  = [5,1,2,3,4,6,7]

arr = sorted(arr)
arr
length =len(arr)

if length%2 ==0:
    print((arr[length//2 -1] +  arr[length//2]) / 2)

else:
    print(length//2)



3


In [ ]:
7/3, 7//3, 7%3

(2.3333333333333335, 2, 1)

## Lonely Integer

In [ ]:
a =[1,2,3,4,3,2,1]   # all lements but one occur twixe
a


[1, 2, 3, 4, 3, 2, 1]

In [ ]:
def lonelyinteger(a):
    result = 0
    for num in a:
        result ^= num  # XOR each element
        print(result)
    return result
lonelyinteger(a)

1
3
0
4
7
5
4


4

In [ ]:
a = [1,2,3]
a =[1,2,3,4,3,2,1]   # all lements but one occur twixe

# final

def lonelyInter(a):
    sum_of_arr = sum(set(a)) 
    # print(sum_of_arr)

    return sum_of_arr*2 - sum(a)

lonelyInter(a)


4

In [ ]:
5**2

25

## Grading Students



1. If the difference between the grade and the next multiple of 5 is less than 3, round grade up to the next multiple of 5.

2. If the value of grade is less than 38, no rounding occurs


In [ ]:
grades = [4, 73, 67, 38, 33]
def gradingStudents(grades):
    for index, value in enumerate(grades):
        if value>=38:
            remainder =value%5
            if remainder>2:
                grades[index] = value + (5 - remainder)
    return grades
gradingStudents(grades)

[4, 75, 67, 40, 33]

## diagonal difference 


[[1, 2, 3], [4, 5, 6], [9, 8, 9]]

In [ ]:
# O(n**2)

x= [[1, 2 ,3],  
     [4 ,5, 6],
     [9 ,8, 9 ],]

x
columns = len(x[0])
rows = len(x)
# print(columns, rows)


diagonal_difference =0
for r in range(rows):
    for c in range(columns):
        if r==c:
            print(x[r][c], x[r][columns - r-1])
            diagonal_difference+= x[r][c]- x[r][columns - r-1]
            # print("\tf:",r,c, "r:", r, columns - r-1])
        
abs(diagonal_difference)

1 3
5 5
9 9


2

In [ ]:
# O(n)
def diagonal_difference(matrix):
  n = len(matrix)
  primary_sum = sum(matrix[i][i] for i in range(n))
  secondary_sum = sum(matrix[i][n - i - 1] for i in range(n))
  return abs(primary_sum - secondary_sum)

diagonal_difference(x)

2

# Counting sort 1

In [ ]:
arr=[1,1,3,2,1,0]
count = [0] * (len(set(arr)))

arr, count

([1, 1, 3, 2, 1, 0], [0, 0, 0, 0])

In [ ]:
arr=[1,1,3,2,1]
# arr = [63, 54, 17, 78, 43, 70, 32, 97, 16, 94, 74, 18, 60, 61, 35, 83, 13, 56, 75, 52, 70, 12, 24, 37, 17, 0, 16, 64, 34, 81, 82, 24, 69, 2, 30, 61, 83, 37, 97, 16, 70, 53, 0, 61, 12, 17, 97, 67, 33, 30, 49, 70, 11, 40, 67, 94, 84, 60, 35, 58, 19, 81, 16, 14, 68, 46, 42, 81, 75, 87, 13, 84, 33, 34, 14, 96, 7, 59, 17, 98, 79, 47, 71, 75, 8, 27, 73, 66, 64, 12, 29, 35, 80, 78, 80, 6, 5, 24, 49, 82]

count = [0] * (max(arr)+1)
print(max(arr))
for v in arr:
  count[v] += 1

count

3


[0, 3, 1, 1]

In [ ]:
[i for i,v in enumerate(count) for _ in range(v)]

[1, 1, 1, 2, 3]

In [ ]:
count
lst=[]
for i, v in enumerate(count):
    for _ in range(v):
        lst.append(i)
lst

[1, 1, 1, 2, 3]

In [ ]:
# sorted
arr=[1,1,3,2,1]

count = [0] * (max(arr)+1)
print(max(arr))
for v in arr:
  count[v] += 1


[i for i,v in enumerate(count) for _ in range(v)]

3


[1, 1, 1, 2, 3]

# Panagram

In [ ]:
s = "We promptly judged antique ivory buckles for the next prize"
# s =s.split()
result = "pagram" if 26 == len(set("".join(s.lower().split()))) else "non pangram"
result

'pagram'

# Mars exploration
A space explorer's ship crashed on Mars! They send a series of SOS messages to Earth for help. 

'SOSSPSSQSSOR'

In [ ]:
S = "SOSSPSSQSSOR"
target_string = "SOS" * (len(S)//3)
S, target_string

('SOSSPSSQSSOR', 'SOSSOSSOSSOS')

In [ ]:
discrepancy = 0
for current, target in zip(S, target_string):
    if target != current:
        discrepancy+=1
discrepancy

3

In [ ]:
'a'*4

'aaaa'

# Counting Valleys

An avid hiker keeps meticulous records of their hikes. During the last hike that took exactly steps, for every step it was noted if it was an uphill, , or a downhill, step. Hikes always start and end at sea level, and each step up or down represents a

unit change in altitude. We define the following terms:

    A mountain is a sequence of consecutive steps above sea level, starting with a step up from sea level and ending with a step down to sea level.
    A valley is a sequence of consecutive steps below sea level, starting with a step down from sea level and ending with a step up to sea level.

Given the sequence of up and down steps during a hike, find and print the number of valleys walked through. 

In [ ]:
number_of_steps = 8
steps = "UDDDUDUU"

number_of_steps, steps

(8, 'UDDDUDUU')

In [ ]:
number_of_steps = 8
steps = "UDDDUDUU"

number_of_steps, steps
downstate = 0
upstate = 0
in_valley =False
for i, value in enumerate(steps):
    if value == 'U':
        upstate+=1
    if value == 'D':
        downstate+=1


    if downstate > upstate and not in_valley:
        in_valley=True
        print("valley-entered")
    if upstate == downstate and in_valley:

        print("valley-exit")  
        in_valley = False  

    # print(f"upstate:{upstate}, downstate:{downstate}, in_valley: {in_valley}")

    


valley-entered
valley-exit


In [ ]:
number_of_steps = 8
steps = "UDDDUDUU"

number_of_steps, steps
elevation = 0
in_valley =False
for i, value in enumerate(steps):
    if value == 'U':
        elevation +=1
    if value == 'D':
        elevation -=1


    if elevation< 0 and not in_valley:
        in_valley=True
        print("valley-entered")
    if elevation == 0 and in_valley:

        print("valley-exit")  
        in_valley = False  

    # print(f"elevation:{elevation}, in_valley: {in_valley}")

    


elevation:1, in_valley: False
elevation:0, in_valley: False
valley-entered
elevation:-1, in_valley: True
elevation:-2, in_valley: True
elevation:-1, in_valley: True
elevation:-2, in_valley: True
elevation:-1, in_valley: True
valley-exit
elevation:0, in_valley: False


In [ ]:
number_of_steps = 8
steps = "UDDDUDUU"

number_of_steps, steps
elevation = valleys = 0
for i, value in enumerate(steps):
    elevation+= 1 if value == 'U' else -1

    if elevation == 0 and value == 'U':
        valleys += 1

    # print(f"elevation:{elevation}, in_valley: {in_valley}")

    
valleys

1

In [ ]:
def countingValleys(steps, path):
    elevation = valleys = 0
    for step in path:
        elevation += 1 if step == 'U' else -1
        if elevation == 0 and step == 'U':
            valleys += 1
    return valleys

# permuting two arrays

In [ ]:
k = 10
A = [2, 1, 3]
B = [7, 8, 9]

def checking_sum(k, A, B):
    A.sort()
    B.sort(reverse=True)
    for a, b in zip(A,B):
        if a + b < k:
            print(a+b)
            return False
        
    return True


checking_sum(k, A, B)

True

# Subarray Division 2
Two children, Lily and Ron, want to share a chocolate bar. Each of the squares has an integer on it. 

In [ ]:
s= [2,2,1,3,2]
d= 4
m=2

for i in range(len(s) -m +1):
    # print(i)
    # sum loop
    temp_sum =0
    
    for t in range(m):
        temp_sum += s[i+t]
        # print(temp_sum)
    if temp_sum == d:
        print(i, i+m, s[i:i+m])
    

0
0 2 [2, 2]
1
2
2 4 [1, 3]
3


In [ ]:
s= [2,2,1,3,2]
d= 4
m=2
count=0


for i in range(len(s) -m +1):
    current_sum =sum(s[i:i+m])
    if(current_sum) == d: 
        count+=1
        print(s[i:i+m])

count

[2, 2]
[1, 3]


2

In [ ]:
s[0:0+2]

[2, 2]

# Sales by Match

In [ ]:
ar = [10, 20, 20, 10, 10, 30, 50, 10, 20]

color_count={}
pairs =0
for i in ar:
    # print(i)
    if i in color_count:
        color_count[i] +=1
    else:
        color_count[i] =1
print(color_count)
# for count in color_count.values():
#     pairs += count//2
#     print(count//2)
# pairs

{10: 4, 20: 3, 30: 1, 50: 1}


# Migratory Birds

In [ ]:
arr = [1,1,2,2,3,7]

{i:v for i,v in enumerate(arr) if i in }

# for i in 

SyntaxError: invalid syntax (171411469.py, line 3)

In [ ]:
arr = [1,1,2,2,3, 7,7]

count_dict = {}

for value in arr:
    if value in count_dict:
        count_dict[value] += 1
    else:
        count_dict[value] = 1

print(count_dict) 

sorted_items = sorted(count_dict.items(), key= lambda x: x[1], reverse=True)
sorted_items[0][0]

{1: 2, 2: 2, 3: 1, 7: 2}


1

In [ ]:
{k:v for k,v in sorted_items}

{1: 2, 2: 2, 7: 2, 3: 1}

In [ ]:
arr = [1,1,2,2,3, 7,7,7, 6,6,6]

bird_count = {}

for value in arr:
    if value in bird_count:
        bird_count[value] += 1
    else:
        bird_count[value] = 1
        
    # Find the bird type with the maximum count
print(bird_count)
max_count = max(bird_count.values())
print(max_count)

# Initialize a variable to hold the smallest bird type ID with the max count
smallest_bird_id = float('inf')


smallest_bird_id = float('inf')
for bird_id, count in bird_count.items():
    if count == max_count and bird_id<smallest_bird_id:
        smallest_bird_id = bird_id

print(smallest_bird_id)

{1: 2, 2: 2, 3: 1, 7: 3, 6: 3}
3
6


# XOR strings 3

In [ ]:
s = "10101"
t = "00101"


In [ ]:
length = max(len(s), len(t))
length    

5

In [ ]:
[0 if s[i] == t[i] else 1 for i in range(len(s))]

[1, 0, 0, 0, 0]

# Drawing Book
A teacher asks the class to open their books to a page number. A student can either start turning pages from the front of the book or from the back of the book. They always turn pages one at a time. When they open the book, page is always on the right side:


    int n: the number of pages in the book
    int p: the page number to turn to


In [ ]:
def pageCount(n, p):
    from_front = p//2
    from_back = (n//2) - (p//2)
    return min(from_front, from_back)
# Example usage:
n = 6  # Total pages
p = 4  # Target page
print(pageCount(n, p))  # Output: 1


1


# Flipping bits

You will be given a list of 32 bit unsigned integers. Flip all the bits ( and ) and return the result as an unsigned integer. 

In [ ]:
def flipping_bits(n):
    bit32='{:032b}'.format(n)
    print(bit32)

    mutated = "".join(['1' if value=='0' else '0' for value in bit32])
    # print(mutated)

    flipped_numer = ''.join(mutated)
    print(flipped_numer)
    print(int(flipped_numer,2))
    


n = 12
flipping_bits(n)

00000000000000000000000000001100
11111111111111111111111111110011
4294967283


In [ ]:
def flipping_bits(n):
    bit32 = '{:032b}'.format(n)  # Convert to 32-bit binary string
    print(f"Original 32-bit binary: {bit32}")

    # Flip bits using list comprehension
    mutated = ['1' if value == '0' else '0' for value in bit32]
    print(f"Flipped bits: {''.join(mutated)}")

    # Convert back to integer
    flipped_number = int(''.join(mutated), 2) #int(..., 2): Converts the binary string back to an integer. The 2 indicates that the string is in base 2 (binary).
    print(f"Flipped integer: {flipped_number}")

    return flipped_number

# Example usage
n = 2147483647
flipping_bits(n)


Original 32-bit binary: 01111111111111111111111111111111
Flipped bits: 10000000000000000000000000000000
Flipped integer: 2147483648


2147483648

In [ ]:
def flipping_bits(n):
   return n ^ 0xFFFFFFFFF  # 32-BIT

# Example usage
n = 2147483647
flipping_bits(n)


66571993088

In [ ]:
def flipping_bits(n):
    bit32 = '{:032b}'.format(n)  # Convert to 32-bit binary string
    print(f"Original 32-bit binary: {bit32}")

    flipped_number = n ^ 0xFFFFFFFF
    # Flip bits using list comprehension
    
    # Convert back to integer from 
    # flipped_number = int(''.join(mutated), 2) #int(..., 2): Converts the binary string back to an integer. The 2 indicates that the string is in base 2 (binary).
    
    flippedbit32 = '{:032b}'.format(flipped_number)  # Convert to 32-bit binary string
    print(f"Original 32-bit binary: {flippedbit32}")
    print(f"Flipped integer: {flipped_number}")

    return flipped_number

# Example usage
n = 2147083647
flipping_bits(n)


Original 32-bit binary: 01111111111110011110010101111111
Original 32-bit binary: 10000000000001100001101010000000
Flipped integer: 2147883648


2147883648

In [ ]:
2^7

5

In [ ]:
    bit32 = '{:04b}'.format(2)  # Convert to 32-bit binary string
    print(f"Original 32-bit binary: {bit32}")
    bit32 = '{:04b}'.format(7)  # Convert to 32-bit binary string
    print(f"Original 32-bit binary: {bit32}")
    x = 2^7
    bit32 = '{:04b}'.format(x)  # Convert to 32-bit binary string
    print(f"Original 32-bit binary: {bit32}")
     


Original 32-bit binary: 0010
Original 32-bit binary: 0111
Original 32-bit binary: 0101


##### XOR
1. different:1
2. same:0

In [ ]:
#

# swapping variables using bit

In [ ]:
a=5;b= 7
bit32 = '{:08b}'.format(a)  # Convert to 32-bit binary string
print(f"{a} - 8-bit binary: {bit32}")

bit32 = '{:08b}'.format(b)  # Convert to 32-bit binary string
print(f"{b} - 8-bit binary: {bit32}")

a = a^b # a contains a^b and b is unchanged a:a^b
bit32 = '{:08b}'.format(a)  # Convert to 32-bit binary string
print(f"{a} - 8-bit binary: {bit32}")
bit32 = '{:08b}'.format(b)  # Convert to 32-bit binary string
print(f"{b} - 8-bit binary: {bit32}")

b = a^b  # b:(a^b)^b  ; b^b=0

bit32 = '{:08b}'.format(a)  # Convert to 32-bit binary string
print(f"{a} - 8-bit binary: {bit32}")
bit32 = '{:08b}'.format(b)  # Convert to 32-bit binary string
print(f"{b} - 8-bit binary: {bit32}")

a = a ^ b # (a^(a^b)) ; a^a=0 = b

bit32 = '{:08b}'.format(a)  # Convert to 32-bit binary string
print(f"{a} - 8-bit binary: {bit32}")
bit32 = '{:08b}'.format(b)  # Convert to 32-bit binary string
print(f"{b} - 8-bit binary: {bit32}")


5 - 8-bit binary: 00000101
7 - 8-bit binary: 00000111
2 - 8-bit binary: 00000010
7 - 8-bit binary: 00000111
2 - 8-bit binary: 00000010
5 - 8-bit binary: 00000101
7 - 8-bit binary: 00000111
5 - 8-bit binary: 00000101


In [ ]:
a, b = 5,7 

a = a ^ b
b = a^b # a^ a ^b
a = a^b # a^b ^b 
a,b

(7, 5)

In [ ]:
a=5
b=7
x = a^b
x^b

5

Original value: 7
After a ^ a: 0 (Binary: 00000000)
After a ^ original: 7 (Binary: 00000111)


In [ ]:
# Initial value
a = 7
print(f"Original value: {a} (Binary: {a:08b})")

# XOR with itself results in 0
a = a ^ a
print(f"After a ^ a: {a} (Binary: {a:08b})")

# XOR with the original number restores the original value
a = a ^ 7
print(f"After a ^ 7: {a} (Binary: {a:08b})")


Original value: 7 (Binary: 00000111)
After a ^ a: 0 (Binary: 00000000)
After a ^ 7: 7 (Binary: 00000111)


# leetcode reverse bits

In [ ]:
def reverseBits( n):
    bit32 = '{:032b}'.format(n)  # Convert to 32-bit binary string
    
    # Flip bits using list comprehension
    reversed_bits = bit32[::-1]

    flipped_number = int(''.join(reversed_bits), 2) #int(..., 2): Converts the binary string back to an integer. 
    
    return flipped_number

reverseBits(43261596)

964176192

# 136 single number

In [ ]:
n = [4,1,2,2,1]
x=0
for i in n:
    x= x^i
    print(f"{i} in bits: {i:08b} | {x} in bits: {x:08b}")

print(x)

4 in bits: 00000100 | 4 in bits: 00000100
1 in bits: 00000001 | 5 in bits: 00000101
2 in bits: 00000010 | 7 in bits: 00000111
2 in bits: 00000010 | 5 in bits: 00000101
1 in bits: 00000001 | 4 in bits: 00000100
4


In [ ]:
# 
from functools import reduce

class Solution:
    def singleNumber(self, nums):
        """
        Find the single number in a list where every other number appears exactly twice.
        
        :type nums: List[int]
        :rtype: int
        """
        return reduce(lambda x, y: x ^ y, nums)

### The reduce(fun,seq) function is used to apply a particular function passed in its argument to all of the list elements mentioned in the sequence passed along.This function is defined in “functools” module.

# flipping matrix

In [1]:
import torch
torch.arange(1,5).reshape(2,2)
sample_input = torch.tensor([[112,42, 83, 119],[56, 125, 56, 49],[15,78,101,43],[62,98,114,108]])
sample_input

tensor([[112,  42,  83, 119],
        [ 56, 125,  56,  49],
        [ 15,  78, 101,  43],
        [ 62,  98, 114, 108]])

In [2]:
# for i in range(len(sample_input)):
#     # print(f"{sample_input[i,:]}: {sum(sample_input[i,:])}  ; {sample_input[:,i]}: {sum(sample_input[:,i])}")
#     print(f"{sample_input[i,:]}: {torch.max(sample_input[i,:])}  ; {sample_input[:,i]}: {torch.max(sample_input[:,i])}")
    

In [8]:
sum = 0
length = len(sample_input)

for r in range(length//2):
    for c in range(length//2):

        sum+= max(
            sample_input[r,c],
            sample_input[r, length-1-c],
            sample_input[length - 1- r, c ],
            sample_input[length - 1 -r , length -1 -c]   
            )
        print(sample_input[r,c],"|",
            sample_input[r, length-1-c],"\n",
            sample_input[length - 1- r, c ],"|",
            sample_input[length - 1 -r , length -1 -c],"\t", sum,"\n",
            )
print(sum)

tensor(112) | tensor(119) 
 tensor(62) | tensor(108) 	 tensor(119) 

tensor(42) | tensor(83) 
 tensor(98) | tensor(114) 	 tensor(233) 

tensor(56) | tensor(49) 
 tensor(15) | tensor(43) 	 tensor(289) 

tensor(125) | tensor(56) 
 tensor(78) | tensor(101) 	 tensor(414) 

tensor(414)


To maximize the sum of the `n x n` top-left quadrant of a `2n x 2n` matrix, we need to carefully consider each element `(i, j)` in this quadrant. For each such element, you can potentially place any one of four elements into this position by flipping rows and/or columns. 

### Step-by-Step Explanation:

For each position `(i, j)` in the top-left `n x n` submatrix, the four potential candidates are:
1. **Top-left element**: The element at `(i, j)` itself.
2. **Top-right element**: The element at `(i, 2n-1-j)`. This element is found by flipping the column that contains `(i, j)`.
3. **Bottom-left element**: The element at `(2n-1-i, j)`. This element is found by flipping the row that contains `(i, j)`.
4. **Bottom-right element**: The element at `(2n-1-i, 2n-1-j)`. This element is found by flipping both the row and column that contain `(i, j)`.

![alt text](flipping_matrix.png)





### Example Matrix
Let's take a `4x4` matrix (for `n=2`) as an example:
```plaintext
Matrix:
112  42  83 119
 56 125  56  49
 15  78 101  43
 62  98 114 108
```

### Analyze Each Position in the Top-Left `2x2` Submatrix:

#### Position (0, 0)
- Candidates:
  1. Top-left: `matrix[0][0] = 112`
  2. Top-right: `matrix[0][3] = 119`
  3. Bottom-left: `matrix[3][0] = 62`
  4. Bottom-right: `matrix[3][3] = 108`
- **Maximum Value**: `max(112, 119, 62, 108) = 119`

#### Position (0, 1)
- Candidates:
  1. Top-left: `matrix[0][1] = 42`
  2. Top-right: `matrix[0][2] = 83`
  3. Bottom-left: `matrix[3][1] = 98`
  4. Bottom-right: `matrix[3][2] = 114`
- **Maximum Value**: `max(42, 83, 98, 114) = 114`

#### Position (1, 0)
- Candidates:
  1. Top-left: `matrix[1][0] = 56`
  2. Top-right: `matrix[1][3] = 49`
  3. Bottom-left: `matrix[2][0] = 15`
  4. Bottom-right: `matrix[2][3] = 43`
- **Maximum Value**: `max(56, 49, 15, 43) = 56`

#### Position (1, 1)
- Candidates:
  1. Top-left: `matrix[1][1] = 125`
  2. Top-right: `matrix[1][2] = 56`
  3. Bottom-left: `matrix[2][1] = 78`
  4. Bottom-right: `matrix[2][2] = 101`
- **Maximum Value**: `max(125, 56, 78, 101) = 125`

### Resulting Maximum Top-Left Quadrant
- The top-left quadrant after choosing the maximum possible values:
```plaintext
119  114
 56  125
```

### Final Sum
- The sum of this quadrant is `119 + 114 + 56 + 125 = 414`.

This approach ensures that the sum of the elements in the top-left quadrant of the matrix is maximized after any number of flips.

In [58]:
# sample_output = torch.tensor([[4,2],[1,3]])
# sample_output

tensor([[4, 2],
        [1, 3]])